File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
import java.util.Stack;
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n= heights.length;
        int[] left= new int[n];
         int[] right= new int[n];
         Stack<Integer> st= new Stack<>();
         for (int i=0;i<n;i++){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i]){
                st.pop();
            }
            left[i]=st.isEmpty()?-1:st.peek();
            st.push(i);
         }
         st.clear();
         for (int i=n-1;i>=0;i--){
            while(!st.isEmpty()&& heights[st.peek()]>=heights[i]){
                st.pop();
            }
            right[i]=st.isEmpty()?n:st.peek();
            st.push(i);
         }
        int maxarea=0;
         for (int i=0;i<n;i++){
            int width= right[i]-left[i]-1;
            int area= heights[i]*width;
            maxarea=Math.max(maxarea,area);
         }
return maxarea;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area = 0
        stack = [-1] 
        n = len(heights)

        for i in range(n + 1):
            current_height = heights[i] if i < n else 0

            while stack[-1] != -1 and heights[stack[-1]] >= current_height:
                h = heights[stack.pop()]
                
                L_i = stack[-1]
                R_i = i
                
                width = R_i - L_i - 1
                
                max_area = max(max_area, h * width)
            
            stack.append(i)

        return max_area
        